src/errors-new.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  MatcherLocation,
+  MatcherLocationNormalized,
+  RouteLocation,
+  RouteLocationNormalized,
+} from './types'
+
+// Using string enums because error codes are exposed to developers
+// and number enums could collide with other error codes in runtime
+export enum ErrorTypes {
+  MATCHER_NOT_FOUND = 'MATCHER_NOT_FOUND',
+  NAVIGATION_GUARD_REDIRECT = 'NAVIGATION_GUARD_REDIRECT',
+  NAVIGATION_ABORTED = 'NAVIGATION_ABORTED',
+  NAVIGATION_CANCELLED = 'NAVIGATION_CANCELLED',
+}
+
+interface RouterError extends Error {
+  type: ErrorTypes
+}
+
+interface MatcherError extends RouterError {
+  type: ErrorTypes.MATCHER_NOT_FOUND
+  location: MatcherLocation
+  currentLocation?: MatcherLocationNormalized
+}
+
+interface NavigationError extends RouterError {
+  type:
+    | ErrorTypes.NAVIGATION_GUARD_REDIRECT
+    | ErrorTypes.NAVIGATION_ABORTED
+    | ErrorTypes.NAVIGATION_CANCELLED
+  from: RouteLocationNormalized
+  to: RouteLocation | RouteLocationNormalized
```

Comment:
Only the NavigationRedirect has this limitation. We should have an extra type that extends Navigation Error for that one